### PR TITLE
ofa det

### PR DESCRIPTION
**Pull request by RachelXu7**
1. 添加ofa在det中的使用
2. 添加picodet ofa demo setting yml文件（模型较复杂，需手动设置搜索空间）
3. trainer中稍做适配，未添加训练策略

### CODE REVIEW

File: ppdet/engine/trainer.py [hunk 1]
```diff
@@ -94,7 +94,10 @@ def __init__(self, cfg, mode='train'):
             self.is_loaded_weights = True
 
         #normalize params for deploy
-        self.model.load_meanstd(cfg['TestReader']['sample_transforms'])
+        if cfg.slim_type == 'OFA':
```
Comment (yghstill) [inline ppdet/engine/trainer.py]: normalize params这个OFA不用适配

File: ppdet/slim/ofa.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    param_state_dict = model.state_dict()
+    paddle.save(model.state_dict(),'check_ofa.pdparams')
+
+    slim_cfg = load_config(slim_cfg)
+
+    ofa_model = build_ofa(model, slim_cfg)
+    
+    input_spec = [{
+        "image": paddle.ones(
+            shape=[1, 3, 640, 640], dtype='float32'),
```
Comment (yghstill) [inline ppdet/slim/ofa.py]: 必须固定[1, 3, 640, 640]？不同模型是否需要配置？

File: configs/slim/ofa/ofa_picodet_demo.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                   'backbone.3_1._conv_linear_2._conv', 'backbone.3_1._conv_dw_mv1._conv',
+                   'backbone.3_1._conv_pw_mv1._conv', 'backbone.3_2._conv_linear._conv',
+                   'backbone.3_3._conv_linear._conv', 'backbone.3_4._conv_linear._conv',
+                   'backbone.3_5._conv_linear._conv', 'backbone.3_6._conv_linear._conv',
+                   'backbone.3_7._conv_linear._conv', 'backbone.4_1._conv_linear_1._conv',
+                   'backbone.4_1._conv_linear_2._conv', 'backbone.4_1._conv_dw_mv1._conv',
+                   'backbone.4_1._conv_pw_mv1._conv', 'backbone.4_2._conv_linear._conv',
+                   'backbone.4_3._conv_linear._conv']
+
+    same_search_space: [
```
Comment (yghstill) [inline configs/slim/ofa/ofa_picodet_demo.yml]: 每个list的含义可否加个注释？

**Comment by yghstill** [comment]
TODO：config/picodet下加个OFA的启动说明吧

File: configs/slim/ofa/ofa_picodet_demo.yml (hunk 3 above)
Comment (RachelXu7) [reply configs/slim/ofa/ofa_picodet_demo.yml]: done

File: ppdet/slim/ofa.py (hunk 2 above)
Comment (RachelXu7) [reply ppdet/slim/ofa.py]: 这个不需要，只是random一个input跑一下前向而已。

File: ppdet/engine/trainer.py (hunk 1 above)
Comment (RachelXu7) [reply ppdet/engine/trainer.py]: 此时self.model已经是ofa model了，不是原本的det模型，self.model.model才是det模型，直接用self.model normalize会报错。

File: ppdet/slim/ofa.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from ppdet.core.workspace import load_config, merge_config, create
+from ppdet.utils.checkpoint import load_weight, load_pretrain_weight
+from ppdet.utils.logger import setup_logger
+
+from paddleslim.nas.ofa import OFA, RunConfig, utils
```
Comment (yghstill) [inline ppdet/slim/ofa.py]: 参考quant，lazy import paddleslim，执行时再import

File: ppdet/slim/ofa.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        "scale_factor": paddle.ones(
+            shape=[1, 2], dtype='float32')
+        }]
+
+    ofa_model._clear_search_space(input_spec=input_spec)
+    ofa_model._build_ss = True
+    check_ss = ofa_model._sample_config('expand_ratio', phase=None)
+    ### tokenize the search space
+    ofa_model.tokenize()
+    ### check token map, search cands and search space
```
Comment (yghstill) [inline ppdet/slim/ofa.py]: 注释只保留一个#

File: configs/datasets/coco_detection.yml [hunk 6]
```diff
@@ -4,15 +4,15 @@ num_classes: 80
 TrainDataset:
   !COCODataSet
     image_dir: train2017
-    anno_path: annotations/instances_train2017.json
-    dataset_dir: dataset/coco
+    anno_path: annotations/instances_minitrain2017.json
```
Comment (yghstill) [inline configs/datasets/coco_detection.yml]: configs/datasets/coco_detection.yml 不要修改~

**Review by yghstill — APPROVED** [review]

**Review by lyuwenyu — APPROVED** [review]
